engine/table/src/main/java/io/deephaven/engine/table/impl/AbstractFilterExecution.java

```diff
@@ -232,77 +196,338 @@ public LogOutput append(LogOutput output) {
+                fc.updatePushdownResult(pushdownResult);
+
+                // Do the next round of filtering with the new filter that bubbled up to the current index.
+                executeStatelessFilter(
+                        filterContexts, filterIdx, filterContexts[filterIdx].filter(),
+                        localInput, filterComplete, filterNec);
+            } else {
+                // Leverage push-down results to reduce the chunk filter input.
+                final Consumer<WritableRowSet> localConsumer = (rows) -> {
+                    try (final RowSet ignored = rows; final PushdownResult ignored2 = pushdownResult) {
+                        onFilterComplete.accept(rows.union(pushdownResult.match()));
+                    }
+                };
+
+                // Do the final filtering at this position.
+                final long inputSize = pushdownResult.maybeMatch().size();
+                if (!shouldParallelizeFilter(filter, inputSize)) {
+                    doFilter(filter, pushdownResult.maybeMatch(), 0, inputSize, localConsumer, filterNec);
+                } else {
+                    doFilterParallel(filter, pushdownResult.maybeMatch(), localConsumer, filterNec);
+                }
+            }
+        };
+
+        final RowSet input = localInput.getValue();
+
+        if (fc.pushdownFilterCost() < Long.MAX_VALUE) {
+            // Execute the pushdown filter and return.
+            if (fc.ppm() != null) {
+                fc.ppm().pushdownFilter(sourceTable.getColumnSourceMap(), currentFilter, input,
+                        sourceTable.getRowSet(), usePrev, fc, costCeiling, jobScheduler(), onPushdownComplete,
+                        filterNec);
+            } else {
+                fc.columnSource().pushdownFilter(currentFilter, input, sourceTable.getRowSet(), usePrev,
+                        fc, costCeiling, jobScheduler(), onPushdownComplete, filterNec);
+            }
             return;
         }
 
-        // Iterate serially through the filters. Each filter will successively restrict the input to the next filter,
-        // until we reach the end of the filter chain.
+        if (fc.pushdownResult() != null) {
+            // Leverage push-down results to reduce the chunk filter input.
+            final PushdownResult localPushdownResult = fc.pushdownResult();
+
+            final Consumer<WritableRowSet> localConsumer = (rows) -> {
+                // Do some cleanup on the rowsets and call the consumer.
+                try (final RowSet ignored = rows; final PushdownResult ignored2 = localPushdownResult) {
+                    onFilterComplete.accept(rows.union(localPushdownResult.match()));
+                }
+            };
+
+            localPushdownResult.match().retain(input);
+            localPushdownResult.maybeMatch().retain(input);
+
+            final long inputSize = localPushdownResult.maybeMatch().size();
```

Comment:
If the inputSize is zero, is it worth actually doing doFilter and doFilterParallel calls?